orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy

```diff
@@ -58,4 +71,67 @@ class CreateServerGroupStage extends AbstractDeployStrategyStage {
 
     return tasks
   }
+
+  @Override
+  List<Stage> onFailureStages(@Nonnull Stage stage) {
+    def stageData = stage.mapTo(StageData)
+
+    if (!stageData.rollbackOnFailure) {
+      // rollback on failure is not enabled
+      return []
+    }
+
+    if (!stageData.getServerGroup()) {
+      // did not get far enough to create a new server group
+      log.warn("No server group was created, skipping rollback! (executionId: ${stage.execution.id}, stageId: ${stage.id})")
+
+      return []
+    }
+
+    return [
+      newStage(
+        stage.execution,
+        rollbackClusterStage.type,
+        "Rollback ${stageData.getCluster()}",
+        [
+          "credentials"   : stageData.getCredentials(),
+          "cloudProvider" : stageData.getCloudProvider(),
+          "regions"       : [stageData.getRegion()],
+          "serverGroup"   : stageData.getServerGroup(),
+          "stageTimeoutMs": TimeUnit.MINUTES.toMillis(30) // timebox a rollback to 30 minutes
+        ],
+        stage,
+        SyntheticStageOwner.STAGE_AFTER
+      )
```

Comment:
This should have a parent stage of the thing being rolled back, no?

Reply:
It should be getting that here ... `STAGE_AFTER` of the deploy that failed

Reply:
Oh, is that what the `stage` param before that is?

Reply:
Yup